velox/common/encode/Base64.h

```diff
@@ -71,56 +76,117 @@ class Base64 {
-  /// result to the 'output'. The output must have enough space, e.g. as
-  /// returned by the calculateEncodedSize().
+  /// result to the 'output' using URL encoding. The output must have enough
+  /// space as returned by the calculateEncodedSize().
   static void encodeUrl(const char* data, size_t size, char* output);
 
-  // compatible with www's Base64URL::encode/decode
-  // TODO rename encode_url/decode_url to encodeUrl/encodeUrl.
+  /// Encodes the specified number of characters from the 'data' using URL
+  /// encoding.
   static std::string encodeUrl(const char* data, size_t len);
+
+  /// Encodes the specified IOBuf data using URL encoding.
   static std::string encodeUrl(const folly::IOBuf* data);
+
+  /// Encodes the specified text using URL encoding.
   static std::string encodeUrl(folly::StringPiece text);
+
+  /// Decodes the specified URL encoded payload and writes the result to the
+  /// 'output'.
   static void decodeUrl(
       const std::pair<const char*, int32_t>& payload,
       std::string& output);
+
+  /// Decodes the specified URL encoded text.
   static std::string decodeUrl(folly::StringPiece text);
 
+  /// Decodes the specified number of characters from the 'src' and writes the
+  /// result to the 'dst'.
   static size_t
   decode(const char* src, size_t src_len, char* dst, size_t dst_len);
 
+  /// Decodes the specified number of characters from the 'src' using URL
+  /// encoding and writes the result to the 'dst'.
   static void
   decodeUrl(const char* src, size_t src_len, char* dst, size_t dst_len);
 
-  constexpr static char kBase64Pad = '=';
-
- private:
+  /// Checks if there is padding in encoded data.
   static inline bool isPadded(const char* data, size_t len) {
-    return (len > 0 && data[len - 1] == kBase64Pad);
+    return (len > 0 && data[len - 1] == kPadding);
   }
 
-  static inline size_t countPadding(const char* src, size_t len) {
+  /// Counts the number of padding characters in encoded data.
+  static inline size_t numPadding(const char* src, size_t len) {
     size_t numPadding{0};
-    while (len > 0 && src[len - 1] == kBase64Pad) {
+    while (len > 0 && src[len - 1] == kPadding) {
       numPadding++;
       len--;
     }
-
     return numPadding;
   }
 
-  static uint8_t Base64ReverseLookup(char p, const ReverseIndex& table);
+  /// Performs a reverse lookup in the reverse index to retrieve the original
+  /// index of a character in the base.
+  static uint8_t base64ReverseLookup(char p, const ReverseIndex& reverseIndex);
+
+  /// Searches for a character within a charset up to a certain index.
+  static constexpr bool
+  findCharacterInCharSet(const Charset& charset, uint8_t idx, const char c) {
+    for (uint8_t i = 0; i <= idx; ++i) {
+      if (charset[i] == c) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // Validate the character in charset with ReverseIndex table
+  static constexpr bool checkForwardIndex(
```

Comment:
Let's follow the existing pattern of using inline for small functions and defining them in the cpp file for larger functions.
```
static bool checkForwardIndex( uint8_t idx,
      const Charset& charset,
      const ReverseIndex& reverseIndex);
```

Reply:
https://github.com/facebookincubator/velox/pull/10371#issuecomment-2241517264